File: testcases/benchmark/sorting/merge_sort.cpp
#include <iostream>
#include <vector>

void merge(std::vector<int>& arr, int l, int m, int r) {
    std::vector<int> L(arr.begin() + l, arr.begin() + m + 1);
    std::vector<int> R(arr.begin() + m + 1, arr.begin() + r + 1);
    int i = 0, j = 0, k = l;
    while (i < (int)L.size() && j < (int)R.size()) {
        if (L[i] <= R[j]) { arr[k] = L[i]; i++; }
        else { arr[k] = R[j]; j++; }
        k++;
    }
    while (i < (int)L.size()) { arr[k] = L[i]; i++; k++; }
    while (j < (int)R.size()) { arr[k] = R[j]; j++; k++; }
}

void merge_sort(std::vector<int>& arr, int l, int r) {
    if (l < r) {
        int m = l + (r - l) / 2;
        merge_sort(arr, l, m);
        merge_sort(arr, m + 1, r);
        merge(arr, l, m, r);
    }
}

int main() { return 0; }

File: testcases/benchmark/sorting/merge_sort.rs
fn merge(arr: &mut Vec<i32>, l: usize, m: usize, r: usize) {
    let left: Vec<i32> = arr[l..=m].to_vec();
    let right: Vec<i32> = arr[m+1..=r].to_vec();
    let (mut i, mut j, mut k) = (0, 0, l);
    while i < left.len() && j < right.len() {
        if left[i] <= right[j] {
            arr[k] = left[i];
            i += 1;
        } else {
            arr[k] = right[j];
            j += 1;
        }
        k += 1;
    }
    while i < left.len() { arr[k] = left[i]; i += 1; k += 1; }
    while j < right.len() { arr[k] = right[j]; j += 1; k += 1; }
}

fn merge_sort(arr: &mut Vec<i32>, l: usize, r: usize) {
    if l < r {
        let m = l + (r - l) / 2;
        merge_sort(arr, l, m);
        merge_sort(arr, m + 1, r);
        merge(arr, l, m, r);
    }
}

fn main() {}
